tools/device-listing/device-listing.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+
+namespace Iot.Device.device_listing.Samples
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            string repoRoot = FindRepoRoot(Environment.CurrentDirectory);
+            string devices = Path.Combine(repoRoot, "src", "devices");
+
+            var deviceListing = new StringBuilder();
+
+            foreach (string directory in Directory.EnumerateDirectories(devices))
+            {
+                if (IsIgnoredDevice(directory))
+                {
+                    continue;
+                }
+
+                string readme = Path.Combine(directory, "README.md");
+                if (File.Exists(readme))
+                {
+                    string title = GetTitle(readme);
+                    if (title == null)
+                    {
+                        Console.WriteLine($"Directory `{directory}` contains readme file without title on the first line.");
+                        continue;
+                    }
+
+                    deviceListing.AppendLine($"* [{title}]({GetRelativePathSimple(readme, devices)})");
+                }
+                else
+                {
+                    Console.WriteLine($"Directory `{directory}` does not have a README.md file.");
+                }
+            }
+
+            ReplacePlaceholder(Path.Combine(devices, "README.md"), "devices", deviceListing.ToString());
+        }
+
+        private static string FindRepoRoot(string dir)
+        {
+            if (!string.IsNullOrEmpty(dir))
+            {
+                if (Directory.Exists(Path.Combine(dir, ".git")))
+                {
+                    return dir;
+                }
+                else
+                {
+                    return FindRepoRoot(new DirectoryInfo(dir).Parent.FullName);
```

Comment:
>FindRepoRoot [](start = 27, length = 12)

should we ensure we are not going outside the repo root?

Reply:
I'm not sure what do you mean.. If we don't find ".git" folder we will return null

Reply:
unless you meant that Parent is null when in root dir - this is now fixed

Reply:
you have fixed this in https://github.com/dotnet/iot/pull/337/commits/4c506bb76782bad6db6ceeddb79789bacd19920b :-)